src/ucp/core/ucp_request.c

```diff
@@ -296,3 +299,112 @@ ucs_status_t ucp_request_test(void *request, ucp_tag_recv_info_t *info)
+    ucp_lane_index_t lane     = config->tag.lane;
+    ucp_worker_h worker       = req->send.ep->worker;
+    size_t only_hdr_size      = proto->only_hdr_size;
+    unsigned flag_iov_single  = 1;
+    unsigned force_sw_rndv    = 0;
+    ucp_rsc_index_t rsc_index;
+    unsigned is_iov;
+    size_t zcopy_thresh;
+    ucs_status_t status;
+    size_t length;
+
+    is_iov = UCP_DT_IS_IOV(req->send.datatype);
+    if (ucs_unlikely(is_iov)) {
+        length = ucp_dt_length(req->send.datatype, count, req->send.buffer,
+                               &req->send.state.dt);
+        req->send.state.dt.dt.iov.iovcnt_offset = 0;
+        req->send.state.dt.dt.iov.iov_offset    = 0;
+        req->send.state.dt.dt.iov.iovcnt        = count;
+        flag_iov_single = (count <= config->tag.eager.max_iov);
+
+        if (!flag_iov_single && ucp_ep_is_tag_offload_enabled(config)) {
+            /* Make sure SW RNDV will be used, because tag offload does
+             * not support multi-packet eager protocols. */
+            force_sw_rndv = 1;
+        }
+
+        if (0 == count) {
+            /* disable zcopy */
+            zcopy_thresh = SIZE_MAX;
+        } else if (!config->tag.eager.zcopy_auto_thresh) {
+            /* The user defined threshold or no zcopy enabled */
+            zcopy_thresh = zcopy_thresh_arr[0];
+        } else if (count <= UCP_MAX_IOV) {
+            /* Using pre-calculated thresholds */
+            zcopy_thresh = zcopy_thresh_arr[count - 1];
+        } else {
+            /* Calculate threshold */
+            rsc_index    = config->key.lanes[lane].rsc_index;
+            zcopy_thresh = ucp_ep_config_get_zcopy_auto_thresh(count,
+                               &ucp_ep_md_attr(req->send.ep, lane)->reg_cost,
+                               worker->context,
+                               worker->ifaces[rsc_index].attr.bandwidth);
+        }
+    } else {
+        length       = ucp_contig_dt_length(req->send.datatype, count);
+        zcopy_thresh = count ? zcopy_thresh_arr[0] : SIZE_MAX;
+    }
+    req->send.length = length;
+
+    ucs_trace_req("select request(%p) progress algorithm datatype=%lx buffer=%p "
+                  " length=%zu max_short=%zd rndv_rma_thresh=%zu rndv_am_thresh=%zu "
+                  "zcopy_thresh=%zu",
+                  req, req->send.datatype, req->send.buffer, length, max_short,
+                  rndv_rma_thresh, rndv_am_thresh, zcopy_thresh);
+
+    req->send.state.uct_comp.func = NULL;
+
+    if (((ssize_t)length <= max_short) && !is_iov) {
+        /* short */
+        req->send.uct.func = proto->contig_short;
+        UCS_PROFILE_REQUEST_EVENT(req, "start_contig_short", req->send.length);
+    } else if ((length >= rndv_rma_thresh) || (length >= rndv_am_thresh) ||
+               force_sw_rndv) {
+        /* RMA/AM rendezvous */
+        ucp_request_send_state_set(req, 0, 0, UCP_REQUEST_SEND_STATE_RNDV);
+        status = ucp_tag_send_start_rndv(req);
```

Comment:
Stream API is not support RNDV yet, i will add the comment, need to add `start_rndv` to `proto`.
looks like it's time for https://github.com/openucx/ucx/pull/1436